### DISCUSSION TITLE
Refresh/Re-render traits of all components in canvas

### DISCUSSION THREAD

**Discussion by ahmedderkaoui** [GrapesJS/grapesjs#3469] [Q&A]
Is there a way to re-render or refresh traits (which are customs, not built-in traits) of all the components existing whithin the canvas?

**Comment by iabhiyaan** [comment]
```js
editor.on('component:selected', component => {
 component.set({
  traits: ['name', 'placeholder']
 })

})
````

**Comment by artf** [comment]
mmm what is the point? Your custom traits should already react to component changes (if you created them properly).

**Reply by ahmedderkaoui** [reply]
I have a custom trait defined as below
```
editor.TraitManager.addType('typeinput', {
  let tableform = '';
  if (component.closest('form')) {
       tableform = component.closest('form').getAttributes().tablename;
  }
  // Here i'm creating my trait select options based on the value of 'tableform' which i got from the above DOM element with tag 'form' in the page;
  ..
  ..
}
```
This code of the custom trait `typeinput` is rendered when the component is selected in the Canvas.

Let's suppose we select this component and the custom is rendered correctly and all works fine. My point is when the value of `component.closest('form').getAttributes().tablename;` is changed in the closest 'form' element, I want to re-render the traits of all components that have the custom trait `'typeinput'` in order to refresh the options created in the trait which are based on the value of that attribute.
It's like i want to delete the whole view of the trait and re-render it like it's the first time it's loaded.

**Reply by artf** [reply]
I think you just need to move that code from `createInput` to `onUpdate` and update your options there
